hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicatorTest.java

```diff
@@ -437,4 +444,275 @@ public void testReplicateFailsToSubmitTask() throws IOException {
+    // The expected batch should include the special mutation.
+    Map<ByteRange, List<Cell>> expectedBatchOfWal = new HashMap<>();
+    expectedBatchOfWal.put(new SimpleByteRange(getRowKey(1)), walEntry1s.getCells());
+
+    // Everything succeeds.
+    when(mockExecutorService.submit((Callable<Object>) any()))
+        .thenReturn(CompletableFuture.completedFuture(true));
+
+    // Trigger replication
+    assertTrue(replicator.replicate(walsToReplicate));
+
+    // called during constructor
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_DELETES_METRICS_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_TIMESTAMP_OVERFLOW_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(PUTS_IN_FUTURE_METRIC_KEY, 0);
+    // Metric reflects that one hbase row mutation was replicated.
+    verify(mockMetricExporter).incCounters(SOURCE_HBASE_REPLICATED, 1);
+
+    verify(mockExecutorService)
+        .submit(
+            new CloudBigtableReplicationTask(
+                TABLE_NAME_STRING, mockConnection, expectedBatchOfWal));
+    Mockito.verifyNoMoreInteractions(mockMetricExporter, mockExecutorService);
+    Mockito.verifyNoInteractions(mockConnection);
+  }
+
+  /**
+   * Two-way replication should drop mutations that come from an external replicated source.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testTwoWayReplicationDropsReplicatedMutation() throws IOException {
+    // Create object to test
+    CloudBigtableReplicator replicator = new CloudBigtableReplicator();
+    replicator.start(sharedResources, incompatibleMutationAdapter, 2000, false);
+
+    // Enable two-way replication manually.
+    Configuration replicationConf = new Configuration(false);
+    replicationConf.setBoolean(ENABLE_TWO_WAY_REPLICATION_MODE_KEY, true);
+    replicator.maybeStartTwoWayReplication(replicationConf, mockMetricExporter);
+
+    // Row key 1 WAL comes from CBT and should be filtered out.
+    Cell cell11 = new KeyValue(getRowKey(1), CF1, null, TIMESTAMP, getValue(1));
+    // Special replicated mutation from CBT.
+    Cell cell1s =
+        new KeyValue(
+            getRowKey(1), CF1, DEFAULT_SOURCE_CBT_QUALIFIER.getBytes(), 0l, KeyValue.Type.Delete);
+
+    // Create replicator input WAL. Note that we don't have any expected replicator output.
+    BigtableWALEntry walEntry1 =
+        new BigtableWALEntry(TIMESTAMP, Arrays.asList(cell11, cell1s), TABLE_NAME_STRING);
+    Map<String, List<BigtableWALEntry>> walsToReplicate = new HashMap<>();
+    walsToReplicate.put(TABLE_NAME_STRING, Arrays.asList(walEntry1));
+
+    // Everything succeeds.
+    when(mockExecutorService.submit((Callable<Object>) any()))
+        .thenReturn(CompletableFuture.completedFuture(true));
```

Comment:
where are you validating that the WAL was dropped?

Reply:
Added explicit noInteractions(mockExecutorService.